File: libs/eave-stdlib-ts/src/transformer-ai/token-counter.ts
import { encoding_for_model } from "tiktoken";
import { OpenAIModel } from "./models.js";

/**
 * Cost per 1k prompt tokens by model.
 * https://openai.com/pricing
 * @param model
 * @returns float price in USD
 */
function inputTokenCost(model: OpenAIModel): number {
  switch (model) {
    case OpenAIModel.GPT_35_TURBO:
      return 0.0015;
    case OpenAIModel.GPT_35_TURBO_16K:
      return 0.003;
    case OpenAIModel.GPT4:
      return 0.03;
    default:
      // using 'never' typed constant here to typecheck that a case hasnt been missed
      // eslint-disable-next-line no-case-declarations
      const missedCase: never = model;
      throw new Error(`Model ${missedCase} not found to compute token cost!`);
  }
}

/**
 * Cost per 1k response tokens by model.
 * https://openai.com/pricing
 * @param model
 * @returns float price in USD
 */
function outputTokenCost(model: OpenAIModel): number {
  switch (model) {
    case OpenAIModel.GPT_35_TURBO:
      return 0.002;
    case OpenAIModel.GPT_35_TURBO_16K:
      return 0.004;
    case OpenAIModel.GPT4:
      return 0.06;
    default:
      // using 'never' typed constant here to typecheck that a case hasnt been missed
      // eslint-disable-next-line no-case-declarations
      const missedCase: never = model;
      throw new Error(`Model ${missedCase} not found to compute token cost!`);
  }
}

export function tokenCount(data: string, model: OpenAIModel): number {
  const encoder = encoding_for_model(model);
  const count = encoder.encode(data).length;
  encoder.free();
  return count;
}

/**
 * Cost of an input prompt to the OpenAI api
 * @param inputTokenCount number of tokens input to model
 * @param model
 * @returns float price in USD
 */
export function calculatePromptCostUSD(inputTokenCount: number, model: OpenAIModel): number {
  return (inputTokenCount / 1000) * inputTokenCost(model);
}

/**
 * Cost of an output response from the OpenAI api
 * @param outputTokenCount number of tokens output from model
 * @param model
 * @returns float price in USD
 */
export function calculateResponseCostUSD(outputTokenCount: number, model: OpenAIModel): number {
  return (outputTokenCount / 1000) * outputTokenCost(model);
}

File: libs/eave-stdlib-py/src/eave/stdlib/transformer_ai/token_counter.py
import tiktoken

from .models import OpenAIModel


def token_count(data: str, model: OpenAIModel) -> int:
    encoder = tiktoken.encoding_for_model(model)
    return len(encoder.encode(data))


def _input_token_cost(model: OpenAIModel) -> float:
    """
    Cost per 1k prompt tokens by model.
    https://openai.com/pricing

    returns float price in USD
    """
    match (model):
        case OpenAIModel.GPT_35_TURBO:
            return 0.0015
        case OpenAIModel.GPT_35_TURBO_16K:
            return 0.003
        case OpenAIModel.GPT4:
            return 0.03


def _output_token_cost(model: OpenAIModel) -> float:
    """
    Cost per 1k prompt tokens by model.
    https://openai.com/pricing

    returns float price in USD
    """
    match (model):
        case OpenAIModel.GPT_35_TURBO:
            return 0.002
        case OpenAIModel.GPT_35_TURBO_16K:
            return 0.004
        case OpenAIModel.GPT4:
            return 0.06


def calculate_prompt_cost_usd(input_token_count: int, model: OpenAIModel) -> float:
    """
    Cost of an input prompt to the OpenAI api
    returns float price in USD
    """
    return (input_token_count / 1000) * _input_token_cost(model)


def calculate_response_cost_usd(output_token_count: int, model: OpenAIModel) -> float:
    """
    Cost of an output response to the OpenAI api
    returns float price in USD
    """
    return (output_token_count / 1000) * _output_token_cost(model)
